neo/NeoSystem.cs

```diff
@@ -56,9 +56,10 @@ public void StartNode(int port = 0, int ws_port = 0)
             });
         }
 
-        public void StartRpc(IPAddress bindAddress, int port, Wallet wallet = null, string sslCert = null, string password = null, string[] trustedAuthorities = null)
+        public void StartRpc(IPAddress bindAddress, int port, Wallet wallet = null, string sslCert = null, string password = null, 
+            string[] trustedAuthorities = null, long maxGasInvoke = 99000000000L)
```

Comment:
Since we allow users to set `maxGasInvoke` through the configuration file, should we set the default value to 10gas as before?

Reply:
Some users pointed out to me that they may want to be able to test the operation of deploy contract before performing it; so that was the reason for allowing the 1000 GAS default. Also some people may use the RPC call to make a number of aggregated get calls from contracts. It was brought to my attention that NEON wallet does this for retrieving the balance of NEP5 tokens and currently consumes upward of 60 GAS in an RPC call. I think 1000 GAS is sufficient for now without leaving the door open for too much abuse.

Reply:
Setting the default value to 10gas allows developers who want to test whether a transaction can be completed for free without additional settings on the configuration file.

Reply:
Yeah, but it returns the amount of gas consumed so that can be achieved with the default of 1000 GAS also.

Reply:
Before 2.9.0 the rpcserver did not have a limit at all. Having 1000 as a limit is closer to the previous behavior than 10 GAS. If the idea is to be closer to what existing systems rely on this is better than 10 GAS as a default. Few production systems had adopted 2.9.0 up to the point at which I created this PR to address the issue that this API was now limiting to only 10 GAS.

Reply:
See https://github.com/neo-project/neo/issues/396

Reply:
I guess I see what you mean, since you mean systems that were not passing the testMode flag as true before.

Reply:
Hmm. It is a good observation. I guess having the default be 10 GAS may be better then, but the RpcServer should override it to 1000 GAS

Reply:
I can make that adjustment as well.

Reply:
Ok, I take that back, the default, for ApplicationEngine.Run is still 10 GAS. This change is only for the RpcServer, which preciously called with test mode set true. So there is no adjustment needed. This is ready to be merged IMO.

Reply:
Perhaps calls using so much GAS is too big of a potential DDOS for public nodes, but if we go back to 10 GAS it will break NEON wallet if they use 2.9.0 nodes. I guess they can always change their wallet. They should probably use an explorer to get token balances instead of a massive script.

Reply:
Perhaps calls using so much GAS is too big of a potential DDOS for public nodes, but if we go back to 10 GAS it will break NEON wallet if they use 2.9.0 nodes. I guess they can always change their wallet. They should probably use an explorer to get token balances instead of a massive script.

Reply:
It would be ideal if NEON could rely on the nodes themselves for balance information that way we dont have to rely on another service/app for core functionality

Reply:
It’s inefficent to get balance on every token for an address. The default nodes don’t track balances of NEP5 assets. You can write Plugin that does and try to get nodes to run it if you want. I’m going to switch this back to defaulting to 10 GAS. It’s too big of a burden on seed nodes to allow calls with large GAS consumption.

Reply:
I looked and the neon call was taking sometimes multiple seconds to get all the balances.

Reply:
I will wait for you to change and continue to review.

Reply:
Ok, look again. :-)

Reply:
@comountainclimber I agree with you. What is efficient way to get balances? Duplicate chain data to off-chain database or counting on third parties? Balances are not only numbers but money. I'd rather use stable data from nodes.

Reply:
As you know, if you use rpc to call a NEP5 asset to get a balance for an address, that is not expensive. What is expensive is calling it to detect changes in balances. You can use a third party service or generate your own for helping with detecting balance changes by address; for obtaining the balance for verification purposes for an asset that has a balance or had a balance but an update says no longer has a balance certainly it should not be too expensive to call an rpc data node in that case to validate. The key is not calling it periodically with every possible asset trying to detect when a balance is found on some new token not held. You need another service to handle that for you.

For a service that notifies of NEP5 balance changes held by a particular address that should not be something provided by the data nodes; they should not be required to track structured data in specific contracts’ storage. It is certainly possible to build a service that has stable and reliable balances by address. Asking The default RPC server of data nodes to provide that is not reasonable. Maybe you could get a number of node operators to run a plugin that does it though and have your wallet talk to nodes supporting the plugin only.